padl/utils.py

```diff
@@ -61,3 +63,129 @@ def __getattr__(self, attr):
+        """
+        pos = len(_pd_trace) - 1
+        print(self.default_msg + '\n' + _pd_trace[pos].transform_str)
+
+        while True:
+            try:
+                x = input('> ')
+            except IndexError:
+                continue
+            if x == 'd':
+                pos, msg = self._down_step(pos, _pd_trace)
+            elif x == 'u':
+                pos, msg = self._up_step(pos, _pd_trace)
+            elif x == 'q':
+                self.args = _pd_trace[pos].args
+                self.trans = _pd_trace[pos].transform
+                break
+            elif x == 'w':
+                msg = _pd_trace[pos].code_position
+            elif x == 'i':
+                msg = _pd_trace[pos].args
+            elif x == 'r':
+                self.args = _pd_trace[pos].args
+                self.trans = _pd_trace[pos].transform
+                # This 0 is because the last element carries a problem: when adding the last
+                # element to _pd_trace *Transform.pd_mode* has been already set up to None again.
+                self.repeat(_pd_trace[0].pd_mode, pos)
+            elif x == 'h' or x == 'help':
+                msg = self.default_msg
+            elif x == 't':
+                msg = _pd_trace[pos].transform_str
+            else:
+                i = _pd_trace[pos].args
+                try:
+                    code = compile(x, '', 'single')
+                    exec(code)
+                except Exception as err:
+                    print(err)
+
+            if x in {'d', 'u', 'w', 'i', 'h', 'help', 't'}:
+                print(f'\n{msg}\n')
+
+    def repeat(self, mode: str, pos: int) -> None:
+        """Repeat the execution from the current position *pos* (the same Exception will be
+        produced).
+
+        :param mode: mode ('train', 'eval', 'infer').
+        :param pos: level of the :class:`Transform` we are inspecting.
+        """
+        assert mode in ('train', 'eval', 'infer')
+        if pos == len(_pd_trace) - 1:
+            _pd_trace.clear()
+            self._repeat_entire(mode)
+        _pd_trace.clear()
+        self._repeat_on_stage(mode)
+
+    @staticmethod
+    def _down_step(pos, pd_trace):
+        if pos > 0:
+            pos -= 1
+            return pos, pd_trace[pos].transform_str
+        return pos, 'Reached the bottom.'
+
+    @staticmethod
+    def _up_step(pos, pd_trace):
+        if pos < len(pd_trace) - 1:
+            pos += 1
+            return pos, pd_trace[pos].transform_str
+        return pos, 'Reached top level.'
+
+    def _repeat_entire(self, mode):
+        if mode == 'train':
+            list(self.trans.train_apply(self.args, batch_size=len(self.args), num_workers=0))
+        if mode == 'eval':
+            list(self.trans.eval_apply(self.args, batch_size=len(self.args), num_workers=0))
+        self.trans.infer_apply(self.args)
+
+    def _repeat_on_stage(self, mode):
+        self.trans.pd_call_transform(self.args, mode)
```

Comment:
Just to clarify, `self.trans` is really `pd_preprocess`, `pd_forward`, or `pd_postprocess` when this function is called right?

Reply:
yes